openwisp_controller/config/base/config.py

```diff
@@ -213,9 +213,14 @@ def manage_vpn_clients(cls, action, instance, pk_set, **kwargs):
         This method is called from a django signal (m2m_changed)
         see config.apps.ConfigConfig.connect_signals
         """
-        if action not in ['post_add', 'post_remove']:
+        if action not in ['custom_openwisp', 'post_add', 'post_remove']:
             return
         vpn_client_model = cls.vpn.through
+        # Delete VpnClients when their related template gets removed
+        if action == 'custom_openwisp':
+            vpn_list = instance.templates.filter(type='vpn').values_list('vpn')
+            instance.vpnclient_set.exclude(vpn__in=vpn_list).delete()
```

Comment:
are we sure this also deletes the related certificates? Please add a test which checks this.